mtags/src/main/scala-3/scala/meta/internal/pc/ExtractMethodProvider.scala

```diff
@@ -0,0 +1,163 @@
+final class ExtractMethodProvider(
+    params: OffsetParams,
+    range: l.Range,
+    defnPos: l.Range,
+    driver: InteractiveDriver,
+    config: PresentationCompilerConfig,
+    symbolSearch: SymbolSearch,
+):
+
+  def extractMethod(): List[TextEdit] =
+    val uri = params.uri
+    val filePath = Paths.get(uri)
+    val source = SourceFile.virtual(filePath.toString, params.text)
+    driver.run(uri, source)
+    val unit = driver.currentCtx.run.units.head
+    val pos = driver.sourcePosition(params)
+    val path =
+      Interactive.pathTo(driver.openedTrees(uri), pos)(using driver.currentCtx)
+    given locatedCtx: Context = driver.localContext(params)
+    val indexedCtx = IndexedContext(locatedCtx)
+
+    val printer = MetalsPrinter.standard(
+      indexedCtx,
+      symbolSearch,
+      includeDefaultParam = MetalsPrinter.IncludeDefaultParam.ResolveLater,
+    )
+
+    def valsOnPath(ts: List[tpd.Tree]): List[(TermName, String)] =
+      ts.flatMap(
+        _ match
+          case ValDef(name, tpt, _) => List((name, printer.tpe(tpt.tpe)))
+          case b @ Block(stats, expr) => valsOnPath(stats :+ expr)
+          case t @ Template(_, _, _, _) =>
+            valsOnPath(t.body)
+          case _ => Nil
+      )
+    def extractFromBlock(t: tpd.Tree): List[tpd.Tree] =
+      t match
+        case Block(stats, expr) =>
+          (stats :+ expr).filter(stat => range.encloses(toLSP(stat.sourcePos)))
+        case temp: Template[?] =>
+          temp.body.filter(stat => range.encloses(toLSP(stat.sourcePos)))
+        case other => List(other)
+
+    def localRefs(ts: List[tpd.Tree]): Set[TermName] =
+      val names = Set.newBuilder[TermName]
+      val valDefs = mutable.Set[TermName]()
+      class NamesCollector extends TreeTraverser:
+        override def traverse(t: tpd.Tree)(using Context): Unit =
+          t match
+            case Select(qualifier, name) =>
+              if !valDefs(name.toTermName) then names += name.toTermName
+              traverse(qualifier)
+            case Ident(name) =>
+              if !valDefs(name.toTermName) then names += name.toTermName
+            case vl @ ValDef(name, _, _) =>
+              valDefs += name
+              traverse(vl.rhs)
+            case _ => traverseChildren(t)
+      val traverser = new NamesCollector
+      ts.map(traverser.traverse(_))
+      names.result()
+    end localRefs
+
+    def genName(path: List[tpd.Tree]): String =
+      def defsOnPath(ts: List[tpd.Tree]): Set[String] =
+        ts.flatMap(
+          _ match
+            case DefDef(name, _, _, _) => Seq(name.toString())
+            case b @ Block(stats, expr) => defsOnPath(stats :+ expr)
+            case t @ Template(_, _, _, _) =>
+              defsOnPath(t.body)
+            case _ => Nil
+        ).toSet
+      val usedNames = defsOnPath(path)
+      if !usedNames("newMethod") then "newMethod"
+      else
+        var i = 0
+        while usedNames(s"newMethod$i") do i += 1
+        s"newMethod$i"
```

Comment:
This could be extracted to use for both Scala 2 and 3 similar to how we reuse MEtalsNAmes